bokehjs/src/coffee/core/layout/side_panel.ts

```diff
@@ -0,0 +1,271 @@
+    parallel   : CENTER,
+    normal     : RIGHT,
+    horizontal : RIGHT,
+    vertical   : CENTER,
+  },
+  right: {
+    justified  : CENTER,
+    parallel   : CENTER,
+    normal     : LEFT,
+    horizontal : LEFT,
+    vertical   : CENTER,
+  },
+}
+
+const _align_lookup_negative: {[key in Side]: string} = {
+  above  : RIGHT,
+  below  : LEFT,
+  left   : RIGHT,
+  right  : LEFT,
+}
+
+const _align_lookup_positive: {[key in Side]: string} = {
+  above  : LEFT,
+  below  : RIGHT,
+  left   : RIGHT,
+  right  : LEFT,
+}
+
+export type SizeableView = DOMView & {
+  get_size(): number
+}
+
+export const _view_sizes = new WeakMap<SizeableView, number>()
+export const _view_constraints = new WeakMap<SizeableView, Constraint>()
+
+export function update_panel_constraints(view: SizeableView): void {
+  const s = view.solver
+  const size = view.get_size()
+  let constraint = _view_constraints.get(view)
+
+  if (constraint != null && s.has_constraint(constraint)) {
+    if (_view_sizes.get(view) === size)
+      return
+    s.remove_constraint(constraint)
+  }
+
+  constraint = GE((view.model as any).panel._size, -size)
+  s.add_constraint(constraint)
+
+  _view_sizes.set(view, size)
+  _view_constraints.set(view, constraint)
+}
+
+export class SidePanel extends LayoutCanvas {
+
+  side: Side
+
+  protected _dim: 0 | 1
+  protected _normals: [number, number]
+  protected _size: Variable
+
+  toString(): string {
+    return `${this.type}(${this.id}, ${this.side})`
+  }
+
+  initialize(attrs: any, options: any): void {
+    super.initialize(attrs, options)
+    switch(this.side) {
+      case "above":
+        this._dim = 0
+        this._normals = [0, -1]
+        this._size = this._height
+        break;
+      case "below":
+        this._dim = 0
+        this._normals = [0, 1]
+        this._size = this._height
+        break;
+      case "left":
+        this._dim = 1
+        this._normals = [-1, 0]
+        this._size = this._width
+        break;
+      case "right":
+        this._dim = 1
+        this._normals = [1, 0]
+        this._size = this._width
+        break;
+      default:
+        logger.error(`unrecognized side: '${this.side}'`)
+    }
+  }
+
+  get dimension(): 0 | 1 {
+    return this._dim
+  }
+
+  get normals(): [number, number] {
+    return this._normals
+  }
+
+  get is_horizontal(): boolean {
+    return this.side == "above" || this.side == "below"
+  }
+
+  get is_vertical(): boolean {
+    return this.side == "left" || this.side == "right"
+  }
+
+  apply_label_text_heuristics(ctx: CanvasRenderingContext2D, orient: TextOrient | number): void {
+    const side = this.side
+
+    let baseline: string
+    let align: string
+
+    if (isString(orient)) {
+      baseline = _baseline_lookup[side][orient]
+      align = _align_lookup[side][orient]
+    } else {
+      if (orient === 0) {
+        baseline = "whatever" // XXX: _baseline_lookup[side][orient]
+        align = "whatever" // XXX: _align_lookup[side][orient]
```

Comment:
@bryevdv, this is something to review. The original code (in comments) doesn't make much sense.

Reply:
It's been this way (which is broken, I think) as far back as 0.5 at least. Here is what should happen when angle is 0: the labels should be incrementally different from when orient==0.1 So this branch should do whatever makes that happen. It will probably take a nested if-block, since things are probably different based on the side.